### DISCUSSION TITLE
Avalonia compared with WPF

### DISCUSSION THREAD

**Discussion by arnirichard** [AvaloniaUI/Avalonia#10637] [Q&A]
I made 2 identically functioning projects in WPF and Avalonia

https://github.com/arnirichard/jpeg_avalonia_wpf

In short, users can click a point on the image and see how jpeg will compress 8x8 pixel block and display multiple WriteableBitmap blocks explaining the process.

There are some discrepancies between WPF and Avalonia

I am setting ColumnDefinition Width="*" (MainWindow) but in Avalonia the columns will not be equally sized when there are overflowing TextBlocks. This is unexpected and causes columns to have chain-effect and resizing each other resulting in infinite cycles.

I have actually fixed that problem with the infinite cycles, and of course setting the TextBlock to wrap will fix it.

To reproduce the unequal column sizes start project JpegAvalonia and reduce the window size to cause TextBlock overflow.

But it can result in undefined behaviour if the column widths can be affected by children (when they should be equal). This can cause chain-effect and infinite cycles in the case when 2 columns fight for space.

Clicking the image will take much longer for Avalonia to refresh compared with WPF. This is caused by the TextBlocks (64 per block) which show the values (numbers). WPF seems to perform better.

To reproduce run projects JpegWpf and JpegAvalonia and click the image of the bird. Below will show how long time it takes to call the PropertyChanged and RaisePropertyChanged in MainWindowViewModel.cs. It takes about 4 times longer for Avalonia (both run in Release/Optimized mode). I wonder why. Commenting out the TextBlocks insertion into Canvas will make this much faster. I wonder if there is some trick to render these values faster. At least I do now know about any library that can draw text on WriteableBitmap.

**Comment by Gillibald** [comment]
You should target Avalonia 11 instead of Avalonia 0.10 to get better results

**Comment by arnirichard** [comment]
Thanks, I made the following changes to my project but when I run it the window shows nothing.

Any idea what is the problem?

![image](https://user-images.githubusercontent.com/95013472/224539614-26479a3e-38b3-4567-a6d8-d88f17e25e34.png)

**Reply by Gillibald** [reply]
You don't import any theme

**Reply by Gillibald** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/master/samples/Sandbox/App.axaml#L6

Make sure to reference the Avalonia.Themes.Fluent package

**Reply by maxkatz6** [reply]
https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes

**Comment by arnirichard** [comment]
Thanks, works now.

**Comment by arnirichard** [comment]
The difference is less now, it is 3 times slower but was 4 times slower (compared with WPF). The columns are still not of equal size. What do you think of that?

**Reply by timunie** [reply]
Should not be the case imo. Columm size should work similar. 

Are the performance issues in debug mode or release mode? The debugger can have a performance impact.

**Reply by arnirichard** [reply]
I am testing this in Release Optimized mode. The code for both is nearly identical.

**Reply by arnirichard** [reply]
Column sizes are displayed at the bottom. For WPF it is equal.
![image](https://user-images.githubusercontent.com/95013472/224583862-ae210dcf-eaf5-4a1b-8f54-f32de0d40b39.png)

**Reply by timunie** [reply]
really interesting. Thank you for sharing your sample btw. I tried to reproduce in a simpler sample, but wasn't able to. Anyway, I can reproduce with the sample provided and acc. to my understanding the column sizes should be almost equal (beside some rounding issues for sure).

**Reply by grokys** [reply]
The Avalonia `Grid` code was ported from WPF in fact, so I'd be surprised if the difference came from there.

**Reply by rabbitism** [reply]
@grokys Actually only TextBlock affects Grid width

**Reply by arnirichard** [reply]
> @grokys Actually only TextBlock affects Grid width

If you get the latest version of my project and try to run project called JpegAvaloniaAsync then you will see the Grid will use up cpu and the columns will fight each other for space infinitely.
